drivers: cop: add NXP COP driver

alxrey opened the pull request:
Port NXP cop driver to Zephyr.

alxrey committed:
drivers: cop: add NXP cop driver

Port NXP cop driver to Zephyr

Signed-off-by: Alexandre Rey <alx.rey@icloud.com>

dleach02 commented on soc/nxp/mcx/mcxc/soc.c:
```diff
@@ -116,6 +116,22 @@ __weak void clock_init(void)
 
+#if defined(CONFIG_WDOG_INIT)
+void z_arm_watchdog_init(void)
+{
```
This doesn't need any actual code?

decsny commented:
@michal-smola @NeilChen93 as this is on MCXC, can you please help to review this PR

michal-smola commented:
As @alxrey suggests, it would be probably better to use different Kconfig symbol to disable the watchdog. Symbol WDT_DISABLE_AT_BOOT which is defined in watchdog driver could be used. The symbol is also used in the driver test.
If WDT_DISABLE_AT_BOOT is used instead of WDOG_ENABLE_AT_BOOT, the empty z_arm_watchdog_init function definition could be avoided.

alxrey commented:
The thing is, since the watchdog register is write-once and can’t be disabled and re-enabled later, I’m not too comfortable leaving it up to the application to configure it early enough before it triggers a reset. I think the cleanest solution would be to configure the watchdog entirely during system init, without requiring the user to call anything manually.
The problem then is how to let the user define the watchdog settings in that case.

dleach02 commented:
I'll suggest that we don't have too much concern over counting on the application to configure the wdog early enough. Remember that they will be making their custom board and application so using z_arm_watchdog_init() at boot time is there to basically handle this situation.

michal-smola commented:
@alxrey , configuration of wdog could be done in soc.c using soc_late_init_hook function. When the function is called, devices are already initialized and wdog driver can be used for configuration. I believe it is what you want to achieve. I tried to implement it and it works, but I am not sure if it is good practice.  @dleach02 @decsny , could you comment, please?

Second point is, that I was not able to disable wdog using the driver. The reason is that there is a check in mcux_cop_disable function if wdog is enabled, but it can be enabled only by calling mcux_cop_setup function which makes the first write to COPC register. After the first write the second write to disable wdog is ignored. @alxrey , could you try reproduce and fix it, please?

alxrey commented:
> @alxrey , configuration of wdog could be done in soc.c using soc_late_init_hook function. [...]

Sounds like a good idea, + use of `WDT_DISABLE_AT_BOOT` symbol rather than `WDOG_ENABLE_AT_BOOT`.
I wait for the review of @dleach02 and @decsny, and if they agree, I will implement that.

alxrey committed:
drivers: cop: add support of WDT_DISABLE_AT_BOOT

Add support for CONFIG_WDT_DISABLE_AT_BOOT Kconfig option.

Signed-off-by: Alexandre Rey <alx.rey@icloud.com>

alxrey committed:
soc: nxp: replace WDOG_ENABLE_AT_BOOT by WDT_DISABLE_AT_BOOT

Using WDT_DISABLE_AT_BOOT instead of WDOG_ENABLE_AT_BOOT prevents the
definition of z_arm_watchdog_init, which is important because the COP
watchdog configuration register can only be configured once.

Signed-off-by: Alexandre Rey <alx.rey@icloud.com>

alxrey committed:
boards: nxp: disable COP watchdog by default

To preserve the previous logic, the watchdog must be disabled by default
using the dedicated Kconfig option.

Signed-off-by: Alexandre Rey <alx.rey@icloud.com>

dleach02 commented:
@alxrey will you pick this PR back up now that the release freeze is over?

dleach02 commented on boards/nxp/frdm_mcxc444/Kconfig.defconfig:
```diff
@@ -0,0 +1,12 @@
+
+if BOARD_FRDM_MCXC444
+
+config WDT_DISABLE_AT_BOOT
```
I'm not sure this is the right way to set this config. Looking at the code base, there is HAS_WDT_DISABLE_AT_BOOT config that is usually set in the drivers/watchdog/Kconfig.mcux. 

For reference, see: https://github.com/zephyrproject-rtos/zephyr/blob/9aecdd47c77fab1b21235ae2bc564a3d0dcf9769/drivers/watchdog/Kconfig.stm32#L12

Holt-Sun commented on dts/bindings/watchdog/nxp,cop.yaml:
```diff
@@ -0,0 +1,41 @@
+      - 256
+      - 1024
+      - 8192
+      - 65526
```
Typo, 65536

Holt-Sun commented on drivers/watchdog/wdt_mcux_cop.c:
```diff
@@ -0,0 +1,205 @@
+	case 8192:
+		data->cop_config.timeoutCycles = kCOP_2Power5CyclesOr2Power13Cycles;
+		break;
+	case 65526:
```
65536

yvesll commented on drivers/watchdog/wdt_mcux_cop.c:
```diff
@@ -0,0 +1,205 @@
+	}
+
+	data->timeout_valid = false;
+	data->setup_done = true;
```
One concern here, if user did not enable COP and call `wdt_disable` before use it, do we really need to call `COP_Disable(config->base)` here. It may waste the write once chance.
